History & Tagging features + Bug fixes

cjyothika opened the pull request:
- Implement History feature
- Implement Tagging feature
- Implement assigning question Id in the backend
- Fix history service bugs
- Store JWT in cookie and not session storage
- Add timer to matching service
- Add solo room feature
- Add warning to other matched user when changing questions while matched
- Ensure docker image can be built

YZTangent committed:
Add models and configs for history service

YZTangent committed:
Add controllers and routes to history service

YZTangent committed:
Add history service to frontend services

YZTangent committed:
Merge branch 'master' into history

YZTangent committed:
Add history component skeleton to frontend

Currently untested

YZTangent committed:
Merge branch 'master' into history

YZTangent committed:
Update gitignore to ignore frontend build

YZTangent committed:
Add skeleton for history component in the frontend

Commit:
Update Question Id mechanism

Commit:
Add multi-select

Commit:
Add multi-select

cjyothika committed:
Add multiple categories

cjyothika committed:
Add question counter model

cjyothika committed:
UI adjustments for tags and complexity multi-select

cjyothika committed:
Add warnings for invalid user redirection

cjyothika committed:
Add question search bar in collab

cjyothika committed:
Add Question Swap during session

cjyothika committed:
Add more sensitive search

cjyothika committed:
Add a ways for user to view questions

cjyothika committed:
Add ability to edit questionTags and questionCategories

cjyothika committed:
Merge branch 'master' into history

cjyothika committed:
Merge pull request #42 from YZTangent/history

History

cjyothika committed:
Merge branch 'tagging-and-history' into ui-updates

cjyothika committed:
Implement get and save history for each question

cjyothika committed:
Add UI for history-service

cjyothika committed:
Improve home page UI

cjyothika committed:
Fix regressions in matching and history component

cjyothika committed:
Add saved attempt to history in UI on http request success

cjyothika committed:
Add timeout bar

cjyothika committed:
Ask matched user before changing question

cjyothika committed:
Make sure docker compose works

cjyothika committed:
Attempt to fix running timer in matching feature

chia-yh commented on backend/history-service/index.js:
```diff
@@ -0,0 +1,59 @@
+
+
+
+// const cors = require('cors');
```
Should this be removed?

cjyothika replied:
Not sure since this part is @YZTangent 's. Thoughts?

chia-yh replied:
@YZTangent ?

chia-yh commented on backend/history-service/models/history.model.js:
```diff
@@ -0,0 +1,16 @@
+    userId: String,
+    authors: [
+        {
+            user_id: String,
```
Could I ask for clarification on the necessity of both `userId` and `user_id`?

cjyothika replied:
The original implementation only had authors but since there weren't necessarily 2 authors for every attempt, the index of the main author got confused so I just kept the main author's id as userId and authors as a place to store collaborators. Also, userId is easier to search if say I wanted to bring up a specific user's attempts in the history page compared to checking the authors' array that will have 2 authors where the user might be but he wasn't the one who saved the attempt. TLDR userId - user who saved the attempt and authors - collaborators on solution.

cjyothika replied:
I can delete the authors array since there will be at most 1 collaborator but that's just how the implementation is for now.

cjyothika replied:
@YZTangent thoughts?

YZTangent replied:
Since we're saving individually, the the authors arary is redundant, I'll go remove it. I wrote it with the intention of accounting for autosaving which saves the history to both users, but since we're not doing that then its not needed.

chia-yh commented on backend/question-service/routes/question.routes.js:
```diff
@@ -4,7 +4,7 @@ module.exports = app => {
 
   // Create a new question
-  router.post("/auth/", questions.create);
+  router.post("/", questions.create);
```
Why is `auth` removed here? Does it still route to use `auth_request` in the gateway?

chia-yh commented on docker-compose.yml:
```diff
@@ -67,6 +67,16 @@ services:
+    container_name: history-service
+    restart: unless-stopped
+    ports:
+      - 8005:8005
```
@YZTangent could you verify that this is intentional over just exposing like this '"8005"' as port mapping exposes the ports outside of the network which allows bypassing the gateway

YZTangent replied:
This was written before the API gateway was implemented, I can remove this now

chia-yh commented on frontend/src/app/_services/collab.service.ts:
```diff
@@ -87,4 +92,17 @@ export class CollabService implements OnInit, OnDestroy {
+    this.request$.next([0, question]);
+  }
+
+  public getRequest = () => {
```
Could I check that this gracefully handles the case that the other user does not agree

cjyothika replied:
Right now if the other user does not agree, a message is sent to the asking user that his request was rejected.

chia-yh commented on frontend/src/app/main/multi-select/multi-select.component.css:
```diff
@@ -0,0 +1,79 @@
+.container {
```
I seem to recall a mention that this was referenced from somewhere, should we include a link to credit it?

cjyothika replied:
My bad. This component is actually my own from another project. I will delete it since I did not end up using it and imported the multi-select component from the material library.

chia-yh commented on frontend/src/app/matching/matching.component.ts:
```diff
@@ -27,19 +33,44 @@ export class MatchingComponent {
+  }
+
   getMatch(userDetails: any) {
+    // this.countDown = setInterval(() => {
```
Should this be removed?

cjyothika replied:
That is actually the UI timer for one of the assignments. I commented it out because I need Justin's help to check and debug it first.

chia-yh reviewed (changes requested):
Some minor nits, would also like some clarification on some areas but otherwise LGTM

cjyothika committed:
Fix matching service timer bug

GenFusion122 committed:
UI and path tweaks

GenFusion122 committed:
Minor tweaks

cjyothika committed:
Merge pull request #46 from GenFusion122/pre-submission-checks

Commit:
Add HTML description and remove matching navigation bug

GenFusion122 committed:
Minor UI tweaks

GenFusion122 committed:
Merge pull request #47 from GenFusion122/pre-submission-checks

Minor UI tweaks

GenFusion122 committed:
Minor changes

GenFusion122 committed:
Minor changes

Added a console.log for assignment 5 submission.

GenFusion122 committed:
Minor changes to log messages

cjyothika committed:
Merge pull request #48 from GenFusion122/pre-submission-checks

Minor changes

Commit:
Add error handling to question-creation page

GenFusion122 committed:
Merge pull request #1 from CS3219-AY2324S1/ui-updates

UI updates

Commit:
Add error messages to question creation

GenFusion122 committed:
Merge branch 'ui-updates' into pre-submission-checks

Commit:
Add error messages to question creation

GenFusion122 committed:
Merge branch 'ui-updates' into pre-submission-checks

GenFusion122 committed:
Minor changes and fixes

GenFusion122 committed:
Merge pull request #50 from GenFusion122/pre-submission-checks

Minor changes + bug fixes in history

GenFusion122 committed:
History changes

Changed implementation of authors array in a History object to a userId2 instead since only one other user can author a solution.

GenFusion122 committed:
Merge pull request #51 from GenFusion122/pre-submission-checks

History changes

cjyothika commented on frontend/src/app/collab/collab.component.ts:
```diff
@@ -70,6 +122,54 @@ export class CollabComponent implements OnInit {
+      this.attempts.push(newAttempt)
+    },
+    err => {
+      console.log("An error occurred while saving attempt: " + err.message)
```
I'm not sure if this should be err.message or err.error.message

cjyothika commented on frontend/src/app/home/home.component.html:
```diff
@@ -1,13 +1,59 @@
+                </div>
+              </div>
+            </div>
+    </tr>
```
Should we fix indentation?

GenFusion122 replied:
Fixed

cjyothika commented on frontend/src/app/main/main.component.ts:
```diff
@@ -67,13 +73,25 @@ selector: any;
+        }, (err) => {
+          var errMessage = "An error occurred while adding your question!"
+          if (err.error) {
+            if (err.error.message.includes("duplicate key error")) {
```
Should catch more errors?

GenFusion122 replied:
I think its fine for now honestly, as long as we mention in the report that only titles cannot be duplicated I suppose?

cjyothika commented on frontend/src/app/matching/matching.component.ts:
```diff
@@ -1,72 +1,113 @@
       }
     })
-    this.getQueueLength()
+    // this.getQueueLength()
```
Remove previous queue length artifact comments

GenFusion122 replied:
Removed in most recent push.

cjyothika reviewed (commented):
LGTM. I just have minor questions and suggestions.

GenFusion122 committed:
Add information into README + minor fixes

GenFusion122 committed:
Merge pull request #52 from GenFusion122/pre-submission-checks

Add information into README + minor fixes

GenFusion122 reviewed (approved):
Tested it, it works so LGTM!
